### PR TITLE
Defining priorityClasses for airflow services and workers

### PR DESCRIPTION
**Pull request by amoghrajesh**
As of now, we do not have any priorityClass definitions in our code base. Users will have to make these changes on their own if they opt for priorityClasses for airflow deployment.

Motivation: https://github.com/apache/airflow/pull/31328#discussion_r1200251031

This PR adds support for having priorityClasses for each of the defined services in your airflow deployment as well as for worker pods.

By default the values are set such that the service pods(statsd, pgbouncer, dagProcessor, flower, scheduler, webserver, triggerer) have a higher priority than the airflow workers. That is, the airflow workers will not be able to pre empt the service pods.

 
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by hussein-awala — CHANGES_REQUESTED** [review]
The priority of a Pod indicates its importance relative to other Pods. When a Pod cannot be scheduled, the scheduler attempts to preempt (evict) lower priority Pods in order to facilitate the scheduling of the pending Pod.

IMO, if we want to support creating priority classes in Airflow helm chart, we should let the user choose his own priorities, and I don't think we need multiple template to do that, we can find a way to create multiple priority class using a single template file.

**Comment by amoghrajesh** [comment]
I think the only way that would be possible is to have another helm chart with a priority class defined and override the helm values for it. 

@hussein-awala do you have any alternative thoughts?

**Review by hussein-awala — COMMENTED** [review]
I was wondering if we can add a new section `priorityClasses` in the values files:
```yaml
priorityClasses: []
#  - name: class1
#    preemptionPolicy: PreemptLowerPriority
#    value: 10000
#  - name: class2
#    preemptionPolicy: Never
#    value: 100000
```
 and create a single template file which loop over this list and create all the priority classes

@jedcunningham wdyt?

**Comment by jedcunningham** [comment]
Seems reasonable to me. We do the same with `extraSecrets` already.

**Comment by amoghrajesh** [comment]
Thanks for that pointer. Makes good sense to me too!

**Comment by amoghrajesh** [comment]
@hussein-awala @jedcunningham I re worked the entire PR. Something like this should be a good fix to such an issue

**Comment by amoghrajesh** [comment]
@hussein-awala @jedcunningham following up for a review on this PR. I have re worked as per your comments and it is ready to be reviewed

File: chart/templates/priorityclasses/priority-classes.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+{{- range $e := .Values.priorityClasses }}
+apiVersion: scheduling.k8s.io/v1
+kind: PriorityClass
+metadata:
+  name: {{ .Release.Name }}-{{ $e.name }}
+{{- if $e.preemptionPolicy }}
+preemptionPolicy: {{ $e.preemptionPolicy }}
+{{ else }}
+preemptionPolicy: PreemptLowerPriority
+{{- end }}
```
Comment (hussein-awala) [inline chart/templates/priorityclasses/priority-classes.yaml]:
```suggestion
preemptionPolicy: {{ $e.preemptionPolicy | default PreemptLowerPriority }}
```

File: chart/templates/priorityclasses/priority-classes.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+value: {{ $e.value }}
```
Comment (hussein-awala) [inline chart/templates/priorityclasses/priority-classes.yaml]: You need to use the required function if it's required or a default value if it is not:
```suggestion
value: {{ $e.value | required "value is required" }}
```

**Review by hussein-awala — COMMENTED** [review]
Could you add some tests for this new feature?

File: chart/templates/priorityclasses/priority-classes.yaml (hunk 1 above)
Comment (amoghrajesh) [reply chart/templates/priorityclasses/priority-classes.yaml]: Good suggestion. Accommodating it

File: chart/templates/priorityclasses/priority-classes.yaml (hunk 2 above)
Comment (amoghrajesh) [reply chart/templates/priorityclasses/priority-classes.yaml]: Good point. Added required clause as it is required and makes more sense than defining a default priority

**Comment by amoghrajesh** [comment]
@hussein-awala the unit tests are ready. The PR is ready for review

**Review by hussein-awala — APPROVED** [review]
LGTM

**Review by potiuk — APPROVED** [review]
NICE

**Pull request merged**